### ISSUE TITLE
Establish connection between Agent and cloud

### ISSUE THREAD

**Issue by thiagoftsm** [netdata/netdata#7042]
##### Feature idea summary
We have talked in previous sprints about the necessity to establish a connection between agent and cloud using libmosquitto and MQTT, this issue has been created to create this channel during next sprint.
##### Expected behavior
Agent to communicate with Cloud.

**Comment by stelfrag** [comment]
##### The suggested approach to add MQTT support to the agent (master) is: 

1. Add libmoquitto dependency
The library is readily available in all distributions. There is no need to bring the source code in the project to test this.
1. Create an internal thread that will publish events to the cloud (thread A)
1. Create an internal thread that will subscribe to the cloud and listen for events (thread B)
1. Decide which metrics should be included in step 2
1. Decide actions to be taken by the agent for events received in step 3

##### Thread A (publish to the cloud)
This thread will initiate communication with the cloud. The *libmosqitto* library supports a number of ways to connect to a broker (simple, TLS, over WebSockets with and with TLS) Source: *https://mosquitto.org* 

For testing purposes (POC) a connection to a local broker can be used to quickly test published events. Using port forwarding the tests can be extended to a remote site (broker)

##### Thread B (subscription to channels)
This will provide a way for the cloud to communicate back to the agent. The agent may subscribe to one or more topics and listen for certain directives. 

##### Metrics to send to the cloud
In stage one of the implementation the suggested metrics to be forwarded to the cloud should
be anything that involves alerts from the agent. This will allow the cloud side to gather, consolidate, filter etc alerts from multiple agents. Reports on those events will eventually appear on a dashboard, so initiate further actions (e.g. Raise a different alarm, even physical, if needed)

##### Actions taken by the agent
When an agent subscribes to a topic on cloud, it can receive messages that can be interpreted
as directives (commands). Those include (but not limited) to:

-  Command to set a configuration parameter
This may cause the agent to immediately change a setting or take some other action
-  Request for certain metrics (query the agent's local database)
The query response will use thread A to publish the results to the cloud
_One approach could be that the request itself will define a new unique channel that will be used
to consolidate the query response from all the agents. Of course the agents will need to publish the responses in that channel_
**More discussion is needed on this, especially for security issues**

##### Slave to Master communication
There is no need to establish a slave to master communication using MQTT. All slaves will have their key metrics processed by the master (agent) as provided by the current streaming protocol.

----
#### Action plan
- Phase 1: Thread A (agent talk to the cloud)  **immediate**
- Phase 2: Thread B (cloud talk to the agent)  **To be decided in the future**

Comments please

**Comment by thiagoftsm** [comment]
I agree 100% with your plan, but I have a doubt, you said that it will be necessary to have two threads, one to send and other to receive, is it possible to use only one to establish this communication channel? Case it is, is it better to do different connections?
I understand "simple" as a communication using plain text, case it is the fact, we can discard this, we need to be sure that the channel is safe.
I also agree that alarms are a very good start point, it is easy for us to check and it is an important information that must be on the cloud.

Best regards!

**Comment by stelfrag** [comment]
@thiagoftsm 
I think it will be a more "clean" solution to have two threads :

- You have a "publisher" and a "subscriber", clearly defined 
- The publish thread (thread A) will be implemented first, there is no real reason to come back to it,  after being tested and deployed (you will need two separate connections to the cloud anyway)
Thread A will be "listening" for alerts (endless loop) and pushing to the cloud
Thread B will be waiting for network input (subscription on topics) and take action on that input

You are right, we will not consider the simple communication (non encrypted transmission)

**Comment by Ferroin** [comment]
It would be useful to allow a master to act as a MQTT broker for it's slaves (or even just have a given system on a network act as a broker for the rest of the network).  If done right, this would allow users to avoid the need for two external connections per agent, which in turn:

* Reduces the load on the user's firewalls.
* Reduces the load on the cloud servers.
* Reduces the total amount of data that has to be sent over the user's internet connection. 

Realistically, I'm fairly certain the cloud will have to be the broker for both connections, so I guess this would be more of a proxy than a true broker.

**Comment by thiagoftsm** [comment]
Hi @Ferroin ,

To avoid we raise the number of connections, it is also possible when an agent has the stream enabled, we do not create a connection with the cloud, instead the agent would send all information to the cloud.
I liked your idea that in future steps we can bring everything to MQTT, it is something that it is really necessary to study.

**Comment by Ferroin** [comment]
@thiagoftsm I'm not really talking about shifting everything to MQTT, In fact, I think that's a bad idea as pub/sub protocols are horrendously inefficient for the type of bulk data transfer we're doing when streaming metrics (although, having an MQTT channel accessible from the local network for each alarm would be a neat feature IMO).

My point was more that we realistically _shouldn't_ need each slave to connect to the cloud for either direction.  Aggregating metrics to the master to push to the cloud is a no-brainer, but I think we should also make use of the master to distribute messages from the cloud to each of the slaves (with the MQTT connection between the master and slaves potentially supporting the unencrypted transports).

**Comment by stelfrag** [comment]
@Ferroin I should have made it more clear (than just a one liner) that only the master will connect to the cloud and it makes no sense to have each one of the slaves connect as well (since the master already has the capability to raise alarms) 

> The suggested approach to add MQTT support to the agent (master) is:

Going to phase two (as described) it would make sense to add a broker on the master that will allow slaves to receive messages.

So we would have: 

- Master subscribes to cloud 
- Receives message for its slaves
- Publishes to an internal broker
- All slaves subscribed receive the message

**Comment by stelfrag** [comment]
Starting to do some testing with Phase A

Steps:
- Add the libmosquitto library as a dependency 
- Initialize the library on server startup (connect to local MQTT broker)
- Add code so that raised alerts are published on the connected MQTT server

The connection at this stage will be plain text. External clients will be able to subscribe
to the alert topic.

**Comment by stelfrag** [comment]
##### Additional tests done

- Use the v5 version of the API
- Set dummy callbacks for the connect, publish commands
- Introduce connection parameters (infight messages, qos)
- Test connection retry options
- Add a subscription to a topic callback
Test reception and attempt to execute internal action (e.g. health config reload)

##### Key point
Communication from/to the broker is handled by libmosquitto in a separate thread by executing the *mosquitto_loop_start* function. There is no need to start a thread explicitly  to handle messages to/from the broker

**Comment by stelfrag** [comment]
###### Current work

- Trying to a CA working so a secure connection can be tested
- This is in parallel to the cloud side of the investigation

##### Todo
- Decide on config file and parameters to use when connecting to the cloud
With regards to that I am reviewing #7086 and the linked issued to that

**Comment by cosmix** [comment]
@stelfrag please add conclusions and key results from this spike to the documentation repo. cc @thiagoftsm
